File: Python3/2744_find_maximum_number_of_string_pairs.py
# https://leetcode.cn/problems/find-maximum-number-of-string-pairs/

class Solution:
    def maximumNumberOfStringPairs(self, words):
        """
        Inputs:
            words: list[str]
        Outputs:
            res: int
        """
        res = 0
        s = set()
        for word in words:
            if word[::-1] in s:
                res += 1
            else:
                s.add(word)
        return res

File: Cpp/2744_find_maximum_number_of_string_pairs.cpp
// https://leetcode.cn/problems/find-maximum-number-of-string-pairs/

class Solution {
public:
    int maximumNumberOfStringPairs(vector<string>& words) {
        int res = 0;
        unordered_set<string> s;
        for (auto &word: words) {
            string word_reverse = string(word.rbegin(), word.rend());
            if (s.find(word_reverse) != s.end()) {
                res += 1;
            }
            else {
                s.insert(word);
            }
        }
        return res;
    }
};
